homeassistant/components/unifiprotect/services.py

```diff
@@ -42,93 +38,78 @@
 )
 
 
-def _async_all_ufp_instances(hass: HomeAssistant) -> list[ProtectApiClient]:
+def _async_all_ufp_instances(hass: HomeAssistant) -> list[ProtectData]:
     """All active UFP instances."""
     return [
-        data.api for data in hass.data[DOMAIN].values() if isinstance(data, ProtectData)
+        data for data in hass.data[DOMAIN].values() if isinstance(data, ProtectData)
     ]
 
 
 @callback
-def _async_get_macs_for_device(device_entry: dr.DeviceEntry) -> list[str]:
-    return [
-        _async_unifi_mac_from_hass(cval)
-        for ctype, cval in device_entry.connections
-        if ctype == dr.CONNECTION_NETWORK_MAC
-    ]
-
-
-@callback
-def _async_get_ufp_instances(
-    hass: HomeAssistant, device_id: str
-) -> tuple[dr.DeviceEntry, ProtectApiClient]:
+def _async_get_ufp_instance(hass: HomeAssistant, device_id: str) -> ProtectData:
     device_registry = dr.async_get(hass)
     if not (device_entry := device_registry.async_get(device_id)):
         raise HomeAssistantError(f"No device found for device id: {device_id}")
 
     if device_entry.via_device_id is not None:
-        return _async_get_ufp_instances(hass, device_entry.via_device_id)
+        return _async_get_ufp_instance(hass, device_entry.via_device_id)
 
-    macs = _async_get_macs_for_device(device_entry)
     ufp_instances = [
-        i for i in _async_all_ufp_instances(hass) if i.bootstrap.nvr.mac in macs
+        i
+        for i in _async_all_ufp_instances(hass)
+        if i.async_get_ufp_device_from_device(device_id) is not None
     ]
 
     if not ufp_instances:
         # should not be possible unless user manually enters a bad device ID
         raise HomeAssistantError(  # pragma: no cover
-            f"No UniFi Protect NVR found for device ID: {device_id}"
+            f"No UniFi Protect Config Entry found for device ID: {device_id}"
         )
 
-    return device_entry, ufp_instances[0]
+    return ufp_instances[0]
 
 
 @callback
 def _async_get_protect_from_call(
     hass: HomeAssistant, call: ServiceCall
-) -> list[tuple[dr.DeviceEntry, ProtectApiClient]]:
+) -> tuple[ProtectData, str]:
     referenced = async_extract_referenced_entity_ids(hass, call)
 
-    instances: list[tuple[dr.DeviceEntry, ProtectApiClient]] = []
-    for device_id in referenced.referenced_devices:
-        instances.append(_async_get_ufp_instances(hass, device_id))
+    if len(referenced.referenced_devices) == 0:
+        raise HomeAssistantError("No referenced devices")  # pragma: no cover
```

Comment:
This is unreachable because https://github.com/home-assistant/core/blob/5e8dc22bda5111dc31d91c0bd4383821fec42f83/homeassistant/components/unifiprotect/services.py#L37 will validate that there is at least one device

Reply:
`async_extract_referenced_entity_ids` only adds the device if it resolves to a valid device. This actually goes with the other mentioned one about resolving UFP Devices. The race condition could in theory happen. 

1. User has _something_ that starts one of the services. Service arguments are validated and task/coroutine for the handler is created. 
2. Before that handler is ran, something causes the device to become invalid.
3. `async_extract_referenced_entity_ids` returns no reference devices (or device lookup fails).

The code is not 100% synchronous so there is no promise that a device that has been validated will still exist when you go to use it.

Reply:
Oh nevermind, it looks like it _always_ adds it to `referenced_devices`, but if it does not resolve it _also_ adds it to `missing_devices`